zerver/views/user_settings.py

```diff
@@ -355,6 +371,72 @@ def json_change_settings(
     # TODO: Change the cache flushing strategy to make sure cache
     # does not contain stale objects.
     user_profile = UserProfile.objects.get(id=user_profile.id)
+
+    # Loop over user_profile.property_types
+    request_settings = {
+        setting_name: requested_value
+        for setting_name, requested_value in locals().items()
+        if setting_name in user_profile.property_types and requested_value is not None
+    }
+
+    if target_users is not None:
+        if not user_profile.is_realm_owner:
+            raise JsonableError(
+                _("Only organization admins can change this setting for other users.")
+            )
+
+        if timezone is not None:
+            # timezone is not part of UserBaseSettings.property_types because it's defined
+            # directly on UserProfile, not in the UserBaseSettings abstract base class.
+            # UserBaseSettings contains only preferences shared between UserProfile and
+            # RealmUserDefault (for organization-wide defaults).
+            request_settings["timezone"] = timezone
+
+        for setting_name in request_settings:
+            if (
+                setting_name in UserBaseSettings.SECURITY_SENSITIVE_USER_SETTINGS
+                or setting_name in ["email", "new_password"]
+            ):
+                raise JsonableError(_("Admin cannot change sensitive settings for other users."))
+
+        user_ids = target_users.user_ids
+        group_ids = target_users.group_ids
+        group_members_dict = get_members_and_subgroups_of_groups(set(group_ids))
+        group_member_ids = set()
+        for members_data in group_members_dict.values():
+            group_member_ids.update(members_data.direct_members)
```

Comment:
We have an optimized function designed for this: `get_recursive_group_members_union_for_groups`